### DISCUSSION TITLE
Sensitivity analysis

### DISCUSSION THREAD

**Discussion by ssingh-ipa** [pybamm-team/PyBaMM#3701] [Q&A]
As per the following discussion,
https://github.com/pybamm-team/PyBaMM/issues/1477
the idea of calculating the sensitivity of the model outputs on the model parameters has already been implemented. 

Are there any examples where the usage of this attribute is shown? If not, please help with the following issue:

options = {"cell geometry": "arbitrary", "thermal": "lumped"}
model = pybamm.lithium_ion.DFN(options)
param = pybamm.ParameterValues("Chen2020")
param.update(
    {
        "Cell cooling surface area [m2]": 0.0040192,
        "Cell volume [m3]": 3.676e-04,
        "Total heat transfer coefficient [W.m-2.K-1]": 10.0, 
    }
)
current_interpolant = pybamm.Interpolant(Time_sec, Current, pybamm.t)
param["Current function [A]"] = current_interpolant
solver = pybamm.CasadiSolver(atol=1e-6, rtol=1e-6, mode="safe")
sim = pybamm.Simulation(model, parameter_values=param, solver=solver)
sol = sim.solve()

plt.plot(Time_sec, Voltage_EK1, color='blue', label='Measured Voltage')
sim_voltage = np.array(sol["Terminal voltage [V]"].entries)
Sim_Time = np.array(sol["Time [s]"].entries)
plt.plot(Sim_Time, sim_voltage, color='red', label="Simulated voltage")
plt.xlabel('Time [s]')
plt.ylabel('Voltage [V]')
plt.legend()
plt.show()

solver2=pybamm.CasadiSolver()
print(solver2)
help(solver2.solve)
inputs = {
    "Negative electrode thickness [m]": 6.73e-05,
    "Positive electrode thickness [m]": 5.55e-05,
}
sens = solver2.solve(model, t_eval=Time_sec, inputs=inputs, calculate_sensitivities=True)

--------------------------------------------------------------------------------------------------
Results and error:

<pybamm.solvers.casadi_solver.CasadiSolver object at 0x000001DE9C960130>
Help on method solve in module pybamm.solvers.base_solver:

solve(model, t_eval=None, inputs=None, initial_conditions=None, nproc=None, calculate_sensitivities=False) method of pybamm.solvers.casadi_solver.CasadiSolver instance
    Execute the solver setup and calculate the solution of the model at
    specified times.
    
    Parameters
    ----------
    model : :class:`pybamm.BaseModel`
        The model whose solution to calculate. Must have attributes rhs and
        initial_conditions. All calls to solve must pass in the same model or
        an error is raised
    t_eval : numeric type
        The times (in seconds) at which to compute the solution
    inputs : dict or list, optional
        A dictionary or list of dictionaries describing any input parameters to
        pass to the model when solving
    initial_conditions : :class:`pybamm.Symbol`, optional
        Initial conditions to use when solving the model. If None (default),
        `model.concatenated_initial_conditions` is used. Otherwise, must be a symbol
        of size `len(model.rhs) + len(model.algebraic)`.
    nproc : int, optional
        Number of processes to use when solving for more than one set of input
        parameters. Defaults to value returned by "os.cpu_count()".
    calculate_sensitivites : list of str or bool
        If true, solver calculates sensitivities of all input parameters.
        If only a subset of sensitivities are required, can also pass a
        list of input parameter names
    
    Returns
    -------
    :class:`pybamm.Solution` or list of :class:`pybamm.Solution` objects.
         If type of `inputs` is `list`, return a list of corresponding
         :class:`pybamm.Solution` objects.
    
    Raises
    ------
    :class:`pybamm.ModelError`
        If an empty model is passed (`model.rhs = {}` and `model.algebraic={}` and
        `model.variables = {}`)
    :class:`RuntimeError`
        If multiple calls to `solve` pass in different models

---------------------------------------------------------------------------
DiscretisationError: Spatial method has not been given for variable Negative particle concentration [mol.m-3] with domain ['negative particle']

DiscretisationError: Cannot automatically discretise model, model should be discretised before solving (Spatial method has not been given for variable Negative particle concentration [mol.m-3] with domain ['negative particle'])

The model had no discretization issues until now. The issue persists even after doing the following discretization:
disc = pybamm.Discretisation()
disc.process_model(model)
 

Please refer me to an example where the attribute calculate_sensitivites has been used.

**Comment by martinjrobins** [comment]
You won't be able to calculate the sensitivities with respect to any of the geometical parameters like "Negative electrode thickness [m]", as these are used in the discretisation of the equations and we cannot take the gradient in this case.

**Reply by ssingh-ipa** [reply]
So, any parameters other than geometrical parameters should work, but when I tried the same code with 'Negative electrode charge transfer coefficient,' I encountered the same error. Can you suggest why?

DiscretisationError: Cannot automatically discretise model, model should be discretised before solving (Spatial method has not been given for variable Negative particle concentration [mol.m-3] with domain ['negative particle'])

Are there any other parameters that need to be added to this list of geometrical parameters?
1. Cell cooling surface area [m2]
2. Electrode height [m]
3. Electrode width [m]
4. Negative current collector thickness [m]
5. Negative electrode thickness [m]
6. Negative particle radius [m]
7. Positive current collector thickness [m]
8. Positive electrode thickness [m]
9. Positive particle radius [m]
10. Separator thickness [m]

**Reply by martinjrobins** [reply]
Here is an example of the sensitivities in use: https://github.com/pybamm-team/workshop-notebooks/blob/main/2021-09-workshop/Session%203/sensitivities_and_data_fitting.ipynb

Looking through your code above, I cannot see where you set the input parameters of the model (look under the "input parameters" heading here https://train.oxrse.uk/material/libraries/pybamm/03_parameter_values). The sensitivities are calculated wrt these input parameters.

**Reply by ssingh-ipa** [reply]
In both the provided links, the input parameter is the current function. 
I tried to run the DFN model with this input:

param.update({"Positive electrode double-layer capacity [F.m-2]": "[input]"})
param.process_geometry(geometry)
param.process_model(model)

# set mesh
var = pybamm.standard_spatial_vars
var_pts = {var.x_n: 30, var.x_s: 30, var.x_p: 30, var.r_n: 10, var.r_p: 10}
mesh = pybamm.Mesh(geometry, model.default_submesh_types, var_pts)

# discretise model
disc = pybamm.Discretisation(mesh, model.default_spatial_methods)
disc.process_model(model)

# create solver
t_eval = np.linspace(0, 3600, 100)
solver = pybamm.CasadiSolver(mode="safe", atol=1e-6, rtol=1e-3)
solution = solver.solve(
    model, t_eval,
    inputs={"Positive electrode double-layer capacity [F.m-2]": 0.2},
    calculate_sensitivities=True
)
plt.plot(
    t_eval,
    solution["Terminal voltage [V]"].sensitivities["Positive electrode double-layer capacity [F.m-2]"]
)

Is it possible that the calculate_sensitivities attribute is only valid for functions?

**Reply by martinjrobins** [reply]
did you get an error with the code above? Sensitivities will work for non function parameters. it would be useful if you post the full script, then I can recreated the error locally and help you out.

**Reply by ssingh-ipa** [reply]
Thank you for the links, they were helpful. Here is the code which can be recreated locally. 

model = pybamm.lithium_ion.DFN()
param = pybamm.ParameterValues("Chen2020")
geometry = model.default_geometry
#param.update({"Current function [A]": "[input]"})
param.update({"Negative electrode active material volume fraction": "[input]"})
#param.update({"Positive electrode double-layer capacity [F.m-2]": "[input]"})
param.process_geometry(geometry)
param.process_model(model)

var = pybamm.standard_spatial_vars
var_pts = {var.x_n: 30, var.x_s: 30, var.x_p: 30, var.r_n: 10, var.r_p: 10}
mesh = pybamm.Mesh(geometry, model.default_submesh_types, var_pts)

disc = pybamm.Discretisation(mesh, model.default_spatial_methods)
disc.process_model(model)
t_eval = np.linspace(0, 3600, 100)
solver = pybamm.CasadiSolver(mode="safe", atol=1e-6, rtol=1e-3)
solution = solver.solve(
    model, t_eval,
    #inputs={"Positive electrode double-layer capacity [F.m-2]": 0.19},
    inputs={"Negative electrode active material volume fraction": 0.78},
    calculate_sensitivities=True
)
plt.plot(
    t_eval,
    #solution["Terminal voltage [V]"].sensitivities["Positive electrode double-layer capacity [F.m-2]"]
    solution["Terminal voltage [V]"].sensitivities["Negative electrode active material volume fraction"]
)
plt.xlabel(r'$t$')
plt.ylabel('Sensitivities of Terminal voltage wrt Negative electrode active material volume fraction')
plt.show()

 inputs={"Negative electrode active material volume fraction": 0.78} is working, but I get the following error for inputs={"Positive electrode double-layer capacity [F.m-2]": 0.19}
KeyError: 'Positive electrode double-layer capacity [F.m-2]'

I understand the input can also be a set of parameters. In that case, does pybamm run the sensitivity test in a range, for e.g. -+10% for each and every permutation?

**Reply by martinjrobins** [reply]
Hi @ssingh-ipa, the issue here is that "Positive electrode double-layer capacity [F.m-2]" is not a parameter of your model, you can print the parameters of the model using:

```python
model.print_parameter_info()
````
